### PR TITLE
Gitcoin Grants CLR Graphs - breakdown of contribution volume over time

### PR DESCRIPTION
**Pull request by owocki**
##### Description

Adds a grants graph showing the breakdown of fund source over time to each Gitcoin Grant.  Details at https://twitter.com/owocki/status/1210242826329632769

##### Refers/Fixes

User Research/Intution

##### Testing

on my local
screencapture-localhost-8000-grants-122-austin-griffith-ethereum-rampd-2019-12-30-15_28_46

### CODE REVIEW

**Comment by owocki** [comment]
@danlipert just finished this up.. moving back to review

File: app/assets/v2/js/grants/stats.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  };
+
+  var chart = new google.visualization.ColumnChart(target);
+
+  chart.draw(view, options);
+}
+
+$(window).resize(function() {
+  drawChart();
+
```
Comment (thelostone-mc) [inline app/assets/v2/js/grants/stats.js]: could we remove extra line

File: app/grants/templates/grants/detail/tabs.html [hunk 2]
```diff
@@ -69,6 +73,33 @@ <h4 class="m-auto text-center font-weight-semibold">{% trans "No Activity for th
...
+                <p><span class="box box_1"></span> &nbsp;&nbsp;&nbsp;{% trans "Subscription Billing" %}</p>
+                <p><span class="box box_2"></span> &nbsp;&nbsp;&nbsp;{% trans "New Subscriptions" %}</p>
+                <p><span class="box box_3"></span> &nbsp;&nbsp;&nbsp;{% trans "One-Time Contributions" %}</p>
+                <p><span class="box box_5"></span> &nbsp;&nbsp;&nbsp;{% trans "CLR Matching Funds" %}</p>
+              </div>
+            </div>
+          </div>
+          <script>
+            document.max_graph = {{max_graph}};
+            document.history = {{history|safe}};
```
Comment (thelostone-mc) [inline app/grants/templates/grants/detail/tabs.html]: Define it as a variable?

File: app/assets/v2/js/grants/stats.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    bar: { groupWidth: '75%' },
+    colors: [ '#011f4b', '#03396c', '#005b96', '#b3cde0' ],
+    isStacked: true,
+    backgroundColor: 'transparent',
+    height: 400,
+    width: width,
+    vAxis: { title: 'USD', ticks: ticks, format: 'short', gridlines: { color: 'transparent' } }
+  };
+
+  var chart = new google.visualization.ColumnChart(target);
```
Comment (thelostone-mc) [inline app/assets/v2/js/grants/stats.js]: update all var -> const/let

File: app/assets/v2/js/grants/stats.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  var view = new google.visualization.DataView(data);
+  var width = parseInt($('#grant_stats_graph').width());
+
+  var ticks = [];
+  var increment = 100;
+
+  if (document.max_graph > 1000) {
+    increment = 1000;
+  }
+  if (document.max_graph > 10000) {
```
Comment (thelostone-mc) [inline app/assets/v2/js/grants/stats.js]: else if

File: app/assets/v2/js/grants/stats.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
+google.charts.load('current', { packages: [ 'corechart', 'bar' ]});
+google.charts.setOnLoadCallback(drawChart);
+
+function drawChart() {
+  var target = document.getElementById('grant_chart');
```
Comment (thelostone-mc) [inline app/assets/v2/js/grants/stats.js]: use `$('#grant_chart')` ?

File: app/grants/templates/grants/detail/tabs.html [hunk 6]
```diff
@@ -69,6 +73,33 @@ <h4 class="m-auto text-center font-weight-semibold">{% trans "No Activity for th
...
+            <div class="chart_container chart_container--big">
+              <p class="sub-title font-weight-semibold font-caption mb-1 pt-1">CONTRIBUTION HISTORY OVER TIME</p>
+              <div class="chart_container__content">
+                <div id="grant_chart"></div>
+              </div>
+              <div class="grant-status-list">
+                <p><span class="box box_1"></span> &nbsp;&nbsp;&nbsp;{% trans "Subscription Billing" %}</p>
+                <p><span class="box box_2"></span> &nbsp;&nbsp;&nbsp;{% trans "New Subscriptions" %}</p>
+                <p><span class="box box_3"></span> &nbsp;&nbsp;&nbsp;{% trans "One-Time Contributions" %}</p>
+                <p><span class="box box_5"></span> &nbsp;&nbsp;&nbsp;{% trans "CLR Matching Funds" %}</p>
```
Comment (thelostone-mc) [inline app/grants/templates/grants/detail/tabs.html]: ^ can we remove the `&nbsp;&nbsp;&nbsp;` and use css classes like `mr-3`

File: app/grants/templates/grants/detail/info.html [hunk 7]
```diff
@@ -280,6 +280,7 @@ <h5 class="font-subheader pb-2">{% trans "Gas Settings" %}</h5>
         </div>
       </div>
 
+
```
Comment (thelostone-mc) [inline app/grants/templates/grants/detail/info.html]: remove extra line?

File: app/grants/models.py [hunk 8]
```diff
@@ -272,6 +272,49 @@ def org_profile(self):
             return profiles.first()
         return None
 
+    @property
+    def history_by_month(self):
+        # gets the history of contributions to this grant month over month so they can be shown o grant details
+        # returns [["", "Subscription Billing",  "New Subscriptions", "One-Time Contributions", "CLR Matching Funds"], ["December 2017", 5534, 2011, 0, 0], ["January 2018", 10396, 0 , 0, 0 ], ... for each monnth in which this grant has contribution history];
```
Comment (thelostone-mc) [inline app/grants/models.py]: remove comments?

File: app/grants/models.py [hunk 9]
```diff
@@ -272,6 +272,49 @@ def org_profile(self):
...
+            for contrib in sub.subscription_contribution.filter(success=True):
+                #add all contributions
+                key = contrib.created_on.strftime("%Y/%m")
+                subkey = 'One-Time' 
+                if int(contrib.subscription.num_tx_approved) > 1:
+                    if contrib.is_first_in_sequence:
+                        subkey = 'New-Recurring'
+                    else:
+                        subkey = 'Recurring-Recurring'
+                if contrib.subscription.contributor_profile.handle in ['vs77bb', 'gitcoinbot', 'notscottmoore', 'owocki']:
```
Comment (thelostone-mc) [inline app/grants/models.py]: thoughts on moving this to a constant defined above ?
['vs77bb', 'gitcoinbot', 'notscottmoore', 'owocki']`

**Comment by thelostone-mc** [comment]
@owocki  Also given that folks have pointed out that the right side is empty! Would it make sense to combine this with the activity tab as opposed to being it's own ?

ideally the graph is build form the same data source as activity

**Comment by owocki** [comment]
> @owocki Also given that folks have pointed out that the right side is empty! Would it make sense to combine this with the activity tab as opposed to being it's own ?

i dont follow "right side is empty" - what does that mean?

**Comment by thelostone-mc** [comment]
derp my bad !
So @androolloyd stated along with a bunch of other folks that [the left side ](https://user-images.githubusercontent.com/5358146/71673612-42ae7180-2d9f-11ea-95ce-4583bcfc9696.png) is filled with a lot of information while  [the right side aka tabs](https://user-images.githubusercontent.com/5358146/71673633-5063f700-2d9f-11ea-8baa-c8e5bb260bcb.png) look relatively empty 

hence the ping asking if it made sense to combine the graph with activity or nah

**Comment by owocki** [comment]
ok thanks; im indifferent to whether it should be its own tab or on the activity tab

File: app/grants/models.py (hunk 8 above)
Comment (owocki) [reply app/grants/models.py]: hmm... are you sure? im trying to get in the habit of documenting my code more..

File: app/assets/v2/js/grants/stats.js (hunk 4 above)
Comment (owocki) [reply app/assets/v2/js/grants/stats.js]: thatll change the logical execution of this. i could move line 21 above line 18 + add an else if tho

File: app/grants/templates/grants/detail/tabs.html (hunk 2 above)
Comment (owocki) [reply app/grants/templates/grants/detail/tabs.html]: hmm i was anchoring it off the document to give it global scope so i could reuse in the JS file.. let me see if that'll work still

File: app/assets/v2/js/grants/stats.js (hunk 3 above)
Comment (owocki) [reply app/assets/v2/js/grants/stats.js]: updated

File: app/assets/v2/js/grants/stats.js (hunk 1 above)
Comment (owocki) [reply app/assets/v2/js/grants/stats.js]: done

File: app/grants/models.py (hunk 9 above)
Comment (owocki) [reply app/grants/models.py]: updated

File: app/grants/templates/grants/detail/tabs.html (hunk 6 above)
Comment (owocki) [reply app/grants/templates/grants/detail/tabs.html]: updated

File: app/grants/templates/grants/detail/tabs.html (hunk 2 above)
Comment (owocki) [reply app/grants/templates/grants/detail/tabs.html]: updated

File: app/grants/models.py (hunk 8 above)
Comment (danlipert) [reply app/grants/models.py]: comments are good! I know we have a lot of docstring style comments which I think are overkill and lead to boilerplate copy-and-pasting but I think comments are something we should definitely encourage, especially if we want outside contributors to grok what we are doing more easily and hence be able to effectively dogfood more @thelostone-mc @owocki

File: app/grants/models.py [hunk 10]
```diff
@@ -272,6 +272,50 @@ def org_profile(self):
             return profiles.first()
         return None
 
+    @property
+    def history_by_month(self):
+        # gets the history of contributions to this grant month over month so they can be shown o grant details
+        # returns [["", "Subscription Billing",  "New Subscriptions", "One-Time Contributions", "CLR Matching Funds"], ["December 2017", 5534, 2011, 0, 0], ["January 2018", 10396, 0 , 0, 0 ], ... for each monnth in which this grant has contribution history];
+        CLR_PAYOUT_HANDLES = ['vs77bb', 'gitcoinbot', 'notscottmoore', 'owocki']
+        month_to_contribution_numbers = {}
+        for sub in self.subscriptions.all():
```
Comment (danlipert) [inline app/grants/models.py]: `prefetch_related('subscription_contribution')` will save us a bunch of DB lookups I think. Then filter `success=True` in python like `for contrib in [sc for sc in sub.subscription_contribution if sc.success=True]`. What do you think? Cost some CPU but greatly reduces DB lookups so should be faster to execute.

Comment (owocki) [reply app/grants/models.py]: hows this? i tested locally 3f49ec63a11fabc1f359aea8e4123b83900a6fc6

File: app/assets/v2/js/grants/stats.js (hunk 5 above)
Comment (thelostone-mc) [reply app/assets/v2/js/grants/stats.js]: Ah that's cause we are loading it before jquery 😅 
no worries

File: app/grants/models.py (hunk 8 above)
Comment (thelostone-mc) [reply app/grants/models.py]: alright alright! majority wins :P
the comments stay

**Pull request merged**